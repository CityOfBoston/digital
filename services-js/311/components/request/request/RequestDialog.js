// @flow

import React from 'react';
import Head from 'next/head';
import Router from 'next/router';
import { action, observable, autorun } from 'mobx';
// import ScopedError from 'react-scoped-error-component';
import { observer } from 'mobx-react';
import { fromPromise } from 'mobx-utils';
import { css } from 'glamor';
import type { Context } from 'next';
import type { IPromiseBasedObservable } from 'mobx-utils';

import type { RequestAdditions } from '../../../server/next-handlers';

import type { SubmittedRequest, Service } from '../../../data/types';
import type { AppStore } from '../../../data/store';
import RequestForm from '../../../data/store/RequestForm';
import makeLoopbackGraphql from '../../../data/dao/loopback-graphql';
import type { LoopbackGraphql } from '../../../data/dao/loopback-graphql';
import submitRequest from '../../../data/dao/submit-request';
import loadService from '../../../data/dao/load-service';

import LoadingBuildings from '../../common/LoadingBuildings';
import FormDialog from '../../common/FormDialog';
import SectionHeader from '../../common/SectionHeader';

import { MEDIA_LARGE, CENTERED_DIALOG_STYLE } from '../../style-constants';

import QuestionsPane from './QuestionsPane';
import LocationPopUp from './LocationPopUp';
import ContactPane from './ContactPane';
import SubmitPane from './SubmitPane';
import CaseView from '../../case/CaseView';

export type InitialProps = {|
  stage: 'questions' | 'location' | 'contact' | 'submit',
  service: ?Service,
  description: string,
|};

export type Props = {|
  store: AppStore,
  loopbackGraphql: LoopbackGraphql,
  routeToServiceForm: (code: string, stage: string) => Promise<void>,
  setLocationMapActive: (active: boolean) => void,
  ...InitialProps,
|};

const COMMON_DIALOG_STYLE = {
  transition: 'margin 400ms',
  position: 'relative',
  zIndex: 2,
};

export const CORNER_DIALOG_STYLE = css(COMMON_DIALOG_STYLE, {
  margin: 0,
  [MEDIA_LARGE]: {
    // comedy bottom margin to push this up into the corner, otherwise it would
    // get vertically centered.
    margin: '-0px 70% 90vh 40px',
    minWidth: 450,
  },
});

@observer
export default class RequestDialog extends React.Component {
  props: Props;

  requestForm: RequestForm;
  @observable
  requestSubmission: ?IPromiseBasedObservable<SubmittedRequest> = null;

  caseLookupOnSubmitDisposer: Function;

  // Called by RequestLayout
  static async getInitialProps({
    query,
    req,
    res,
  }: Context<RequestAdditions>): Promise<InitialProps> {
    const { code, description } = query;
    let stage = query.stage;

    const loopbackGraphql = makeLoopbackGraphql(req);

    const service = await loadService(loopbackGraphql, code);

    if (!service && res) {
      res.statusCode = 404;
    }

    stage = stage || 'questions';

    switch (stage) {
      case 'questions':
      case 'location':
      case 'contact':
      case 'submit':
        return {
          service,
          stage,
          description: description || '',
        };
      default:
        throw new Error(`Unknown stage: ${stage}`);
    }
  }

  @action
  componentWillMount() {
    const { stage, description, setLocationMapActive, service } = this.props;

    this.requestForm = new RequestForm(service);
    this.requestForm.description = description;

    setLocationMapActive(stage === 'location');
  }

  componentDidMount() {
    LoadingBuildings.preload();

    this.caseLookupOnSubmitDisposer = autorun('route on case submit', () => {
      if (
        this.requestSubmission &&
        this.requestSubmission.state === 'fulfilled'
      ) {
        const { id } = this.requestSubmission.value;
        // We keep the first URL the same to avoid a page load, but change the
        // "as" to match the URL for the case permalink.
        Router.replace('/request', `/reports/${id}`, {
          shallow: true,
        });
      }
    });
  }

  componentWillUnmount() {
    this.caseLookupOnSubmitDisposer();
  }

  @action
  componentWillReceiveProps(newProps: Props) {
    const { setLocationMapActive, stage } = this.props;

    if (stage !== newProps.stage) {
      setLocationMapActive(newProps.stage === 'location');
    }
  }

  makeNextAfterQuestions(): { fn: () => mixed, isSubmit: boolean } {
    const { requestForm } = this;

    if (requestForm.showLocationPicker) {
      return {
        fn: this.routeToLocation,
        isSubmit: false,
      };
    } else if (requestForm.showContactInfoForm) {
      return {
        fn: this.routeToContact,
        isSubmit: false,
      };
    } else {
      return {
        fn: this.submitRequest,
        isSubmit: true,
      };
    }
  }

  makeNextAfterLocation(): { fn: () => mixed, isSubmit: boolean } {
    const { requestForm } = this;

    if (requestForm.showContactInfoForm) {
      return {
        fn: this.routeToContact,
        isSubmit: false,
      };
    } else {
      return {
        fn: this.submitRequest,
        isSubmit: true,
      };
    }
  }

  routeToLocation = () => {
    const { service, routeToServiceForm } = this.props;
    if (service) {
      routeToServiceForm(service.code, 'location');
    }
  };

  routeToContact = () => {
    const { service, routeToServiceForm } = this.props;
    if (service) {
      routeToServiceForm(service.code, 'contact');
    }
  };

  @action.bound
  submitRequest(): Promise<mixed> {
    const { requestForm } = this;
    const { service, loopbackGraphql, routeToServiceForm } = this.props;
    const {
      description,
      firstName,
      lastName,
      email,
      phone,
      location,
      address,
      addressId,
      questions,
      mediaUrl,
      sendLocation,
      sendContactInfo,
    } = requestForm;

    if (!service) {
      throw new Error('service is null in submitRequest');
    }

    const promise = submitRequest(loopbackGraphql, {
      service,
      description,
      // We only send contact info with an affirmative boolean that's set by
      // the ContactPane form's submit button. That way contact info that was
      // loaded by localStorage won't be pushed to the server if the user didn't
      // see the contact info form or chose "submit without sending contact info"
      firstName: sendContactInfo ? firstName : null,
      lastName: sendContactInfo ? lastName : null,
      email: sendContactInfo ? email : null,
      phone: sendContactInfo ? phone : null,
      location: sendLocation ? location : null,
      address: sendLocation ? address : null,
      addressId: sendLocation ? addressId : null,
      questions,
      mediaUrl,
    }).then(
      v => {
        // upload was successful, so clear out previous request state
        this.requestForm = new RequestForm(service);
        return v;
      },
      err => {
        /* eslint-disable no-underscore-dangle */
        if (window._opbeat) {
          window._opbeat('captureException', err);
          err._sentToOpbeat = true;
        }
        /* eslint-enable no-underscore-dangle */

        // This exception is ultimately "caught" by the requestSubmission fromPromise
        // mobx handler.
        throw err;
      }
    );

    this.requestSubmission = fromPromise(promise);

    routeToServiceForm(service.code, 'submit');

    return promise;
  }

  render() {
    const { stage } = this.props;
    const dialogContainerStyle =
      stage === 'location'
        ? CORNER_DIALOG_STYLE
        : css(COMMON_DIALOG_STYLE, CENTERED_DIALOG_STYLE);

    return (
      <div className={dialogContainerStyle}>
        <Head>
          <title>
            BOS:311 — {this.renderTitle()}
          </title>
        </Head>

        <FormDialog
          narrow={stage === 'contact'}
          noPadding={stage === 'location'}
        >
          {this.renderContent()}
        </FormDialog>
      </div>
    );
  }

  renderTitle() {
    const { requestSubmission } = this;
    const { stage, service } = this.props;

    if (!service) {
      return 'Not Found';
    }

    switch (stage) {
      case 'questions':
        return service.name;
      case 'location':
        return 'Choose location';
      case 'contact':
        return 'Contact information';
      case 'submit':
        if (!requestSubmission) {
          return '';
        }

        switch (requestSubmission.state) {
          case 'pending':
            return 'Submitting…';
          case 'fulfilled':
            return `Success: Case ${requestSubmission.value.id}`;
          case 'rejected':
            return 'Submission error';
          default:
            return '';
        }

      default:
        return '';
    }
  }

  renderContent() {
    const { requestSubmission, requestForm } = this;
    const { store, stage, service } = this.props;

    if (!service) {
      return <SectionHeader>Service not found</SectionHeader>;
    }

    // TODO(finh): Bring back service errors
    // if (currentServiceError) {
    //   return (
    //     <div>
    //       <SectionHeader>Error loading service</SectionHeader>
    //       <ScopedError error={currentServiceError} />
    //     </div>
    //   );
    // }

    switch (stage) {
      case 'questions': {
        const { fn, isSubmit } = this.makeNextAfterQuestions();
        return (
          <QuestionsPane
            store={store}
            requestForm={requestForm}
            serviceName={service.name}
            serviceDescription={service.description}
            nextFunc={fn}
            nextIsSubmit={isSubmit}
          />
        );
      }

      case 'location': {
        const { fn, isSubmit } = this.makeNextAfterLocation();
        return (
          <LocationPopUp
            store={store}
            requestForm={requestForm}
            nextFunc={fn}
            nextIsSubmit={isSubmit}
          />
        );
      }

      case 'contact':
        return (
          <ContactPane
            requestForm={requestForm}
            serviceName={service.name}
            nextFunc={this.submitRequest}
          />
        );

      case 'submit':
        if (!requestSubmission) {
          return '';
        }

        switch (requestSubmission.state) {
          case 'pending':
            return <SubmitPane state="submitting" ui={store.ui} />;
          case 'fulfilled':
            return (
              <CaseView
                store={store}
                request={requestSubmission.value}
                submitted
              />
            );
          case 'rejected':
            return <SubmitPane state="error" error={requestSubmission.value} />;
          default:
            return '';
        }

      default:
        return null;
    }
  }
}
