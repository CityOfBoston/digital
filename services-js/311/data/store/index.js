// @flow

import { observable, computed, useStrict } from 'mobx';
import type { IPromiseBasedObservable } from 'mobx-utils';

import type { Service, ServiceSummary, SubmittedRequest } from '../types';
import Question from './Question';

// MobX will enforce that state changes only happen in action blocks.
useStrict(true);

export class ContactInfo {
  @observable required: boolean = false;
  @observable firstName: string = '';
  @observable lastName: string = '';
  @observable email: string = '';
  @observable phone: string = '';

  // can be false even when required is false to differentiate between submitting
  // and skipping
  @computed
  get requirementsMet(): boolean {
    return !!(this.firstName && this.lastName);
  }
}

export class LocationInfo {
  @observable required: boolean = false;
  @observable address: string = '';
  @observable.shallow location: ?{| lat: number, lng: number |} = null;

  // can be false even when required is false to differentiate between submitting
  // and skipping
  @computed
  get requirementsMet(): boolean {
    return this.address.length > 0;
  }
}

export class AppStore {
  @observable description: string = '';

  @observable contactInfo: ContactInfo = new ContactInfo();
  @observable locationInfo: LocationInfo = new LocationInfo();
  @observable questions: Question[] = [];

  @observable.shallow serviceSummaries: ServiceSummary[];
  serviceCache: Map<string, Service> = observable.shallowMap({});

  apiKeys: {[service: string]: string} = {};
  isPhone: boolean = false;

  @observable requestSubmission: ?IPromiseBasedObservable<SubmittedRequest> = null;

  @observable.ref _currentService: ?Service = null;
  @observable.ref currentServiceError: ?Object = null;

  @computed get currentService(): ?Service {
    return this._currentService;
  }

  set currentService(service: ?Service) {
    if (service === this._currentService) {
      return;
    }

    this._currentService = service;
    this.currentServiceError = null;
    this.questions = [];

    if (service) {
      try {
        this.contactInfo.required = service.contactRequired;
        this.locationInfo.required = service.locationRequired;
        this.questions = Question.buildQuestions(service.attributes);
      } catch (e) {
        this.currentServiceError = e;
      }
    }
  }

  @computed
  get questionRequirementsMet(): boolean {
    return this.questions.filter((q) => !q.required || !q.visible || q.requirementsMet).length === this.questions.length;
  }
}

let browserStore: AppStore;

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 */
export default function getStore(): AppStore {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const store = new AppStore();

  if (process.browser) {
    browserStore = store;
  }

  return store;
}
