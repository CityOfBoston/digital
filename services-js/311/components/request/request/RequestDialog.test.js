// @flow

import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { runInAction } from 'mobx';

import RequestDialog from './RequestDialog';

import type { Service, SubmittedRequest } from '../../../data/types';
import { AppStore } from '../../../data/store';

jest.mock('../../../data/dao/submit-request');
const submitRequest: JestMockFn = (require('../../../data/dao/submit-request'): any)
  .default;

const MOCK_SERVICE: Service = {
  name: 'Cosmic Incursion',
  description: 'Bad things getting in from other universes',
  code: 'CSMCINC',
  contactRequirement: 'REQUIRED',
  locationRequirement: 'VISIBLE',
  attributes: [
    {
      required: false,
      type: 'TEXT',
      code: 'ST-CMTS',
      description: 'Please provide any other relevant information:',
      values: null,
      conditionalValues: null,
      dependencies: null,
    },
    {
      required: false,
      type: 'STRING',
      code: 'INFO-CSIRMV1',
      description:
        '**All cosmic incursion cases should be followed up with a phone call to Alpha Flight.**',
      values: null,
      conditionalValues: null,
      dependencies: null,
    },
    {
      required: true,
      type: 'SINGLEVALUELIST',
      code: 'SR-CSIRMV1',
      description: 'How many dimensions have been breached?',
      values: [
        { key: 'One', name: 'One' },
        { key: 'Two', name: 'Two' },
        { key: 'Three', name: 'Three' },
        { key: 'More than Three', name: 'More than Three' },
      ],
      conditionalValues: [],
      dependencies: null,
    },
  ],
};

const MOCK_ACTIONS = {
  loopbackGraphql: jest.fn(),
  routeToServiceForm: jest.fn(),
  setLocationMapActive: jest.fn(),
};

describe('rendering', () => {
  let store;

  beforeEach(() => {
    store = new AppStore();
  });

  test('questions', () => {
    const component = renderer.create(
      <RequestDialog
        store={store}
        stage="questions"
        service={MOCK_SERVICE}
        description=""
        {...MOCK_ACTIONS}
      />,
      { createNodeMock: () => document.createElement('div') }
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('location', () => {
    const component = renderer.create(
      <RequestDialog
        store={store}
        stage="location"
        service={MOCK_SERVICE}
        description=""
        {...MOCK_ACTIONS}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('contact', () => {
    const component = renderer.create(
      <RequestDialog
        store={store}
        stage="contact"
        service={MOCK_SERVICE}
        description=""
        {...MOCK_ACTIONS}
      />,
      { createNodeMock: () => document.createElement('div') }
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('methods', () => {
  let wrapper;
  let store;
  let requestDialog: RequestDialog;
  let loopbackGraphql;
  const routeToServiceForm = jest.fn();

  beforeEach(() => {
    loopbackGraphql = jest.fn();

    store = new AppStore();

    wrapper = shallow(
      <RequestDialog
        store={store}
        description=""
        stage="submit"
        service={MOCK_SERVICE}
        loopbackGraphql={loopbackGraphql}
        routeToServiceForm={routeToServiceForm}
        setLocationMapActive={jest.fn()}
      />
    );

    requestDialog = wrapper.instance();

    runInAction(() => {
      requestDialog.requestForm.firstName = 'Carol';
      requestDialog.requestForm.address = 'City Hall Plaza, Boston, MA';
    });
  });

  describe('navigation', () => {
    test('nextAfterQuestions', () => {
      requestDialog.routeToLocation();
      expect(routeToServiceForm).toHaveBeenCalledWith(
        MOCK_SERVICE.code,
        'location'
      );
    });

    test('nextAfterLocation', () => {
      requestDialog.routeToContact();
      expect(routeToServiceForm).toHaveBeenCalledWith(
        MOCK_SERVICE.code,
        'contact'
      );
    });
  });

  describe('submitRequest', () => {
    let resolveGraphql;
    let rejectGraphql;

    beforeEach(() => {
      const promise = new Promise((resolve, reject) => {
        resolveGraphql = resolve;
        rejectGraphql = reject;
      });

      submitRequest.mockReturnValue(promise);
    });

    test('success', async () => {
      const submission = requestDialog.submitRequest();

      // rendering loading
      wrapper.update();
      expect(wrapper).toMatchSnapshot();

      const result: SubmittedRequest = {
        id: '17-000000001',
        service: {
          name: 'Cosmic Intervention',
        },
        description: 'I think that Thanos is here',
        status: 'closed',
        statusNotes:
          'Found Thanos. Smashed him into the floor with all of us standing around.',
        location: {
          lat: 42.359927299999995,
          lng: -71.0576853,
        },
        mediaUrl: 'https://pbs.twimg.com/media/C22X9ODXgAABGKS.jpg',
        address: 'City Hall Plaza, Boston, MA 02131',
        requestedAtString: 'March 7, 2017, 12:59 PM',
        updatedAtString: 'April 8, 2017, 12:59 PM',
      };

      expect(submitRequest).toHaveBeenCalledWith(
        loopbackGraphql,
        expect.objectContaining({
          firstName: null,
          address: null,
        })
      );

      resolveGraphql(result);

      await submission;

      // rendering success
      wrapper.update();
      expect(wrapper).toMatchSnapshot();
    });

    test('success with contact info and location', async () => {
      runInAction(() => {
        requestDialog.requestForm.sendContactInfo = true;
        requestDialog.requestForm.sendLocation = true;
      });

      requestDialog.submitRequest();

      expect(submitRequest).toHaveBeenCalledWith(
        loopbackGraphql,
        expect.objectContaining({
          firstName: 'Carol',
          address: 'City Hall Plaza, Boston, MA',
        })
      );
    });

    test('graphql failure', async () => {
      const submission = requestDialog.submitRequest();

      const error: Object = new Error('Error submitting');
      error.errors = [
        { message: 'All required fields were missing' },
        { message: 'Also your location is in Cambridge' },
      ];

      rejectGraphql(error);

      try {
        await submission;
      } catch (e) {
        // expected
      }

      // should be rendering the error here
      wrapper.update();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
