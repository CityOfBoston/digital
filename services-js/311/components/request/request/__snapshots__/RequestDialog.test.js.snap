// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`methods submitRequest graphql failure 1`] = `
<div
  className={
    Object {
      "data-css-1prrpm3": "",
    }
  }
>
  <SideEffect(Head)>
    <title>
      BOS:311 — 
      Submission error
    </title>
  </SideEffect(Head)>
  <FormDialog
    narrow={false}
    noPadding={false}
  >
    <SubmitPane
      error={[Error: Error submitting]}
      state="error"
    />
  </FormDialog>
</div>
`;

exports[`methods submitRequest success 1`] = `
<div
  className={
    Object {
      "data-css-1prrpm3": "",
    }
  }
>
  <SideEffect(Head)>
    <title>
      BOS:311 — 
      Submitting…
    </title>
  </SideEffect(Head)>
  <FormDialog
    narrow={false}
    noPadding={false}
  >
    <SubmitPane
      state="submitting"
      ui={
        Ui {
          "handleResize": [Function],
          "handleScroll": [Function],
        }
      }
    />
  </FormDialog>
</div>
`;

exports[`methods submitRequest success 2`] = `
<div
  className={
    Object {
      "data-css-1prrpm3": "",
    }
  }
>
  <SideEffect(Head)>
    <title>
      BOS:311 — 
      Success: Case 17-000000001
    </title>
  </SideEffect(Head)>
  <FormDialog
    narrow={false}
    noPadding={false}
  >
    <CaseView
      noMap={false}
      request={
        Object {
          "address": "City Hall Plaza, Boston, MA 02131",
          "description": "I think that Thanos is here",
          "id": "17-000000001",
          "images": Array [
            Object {
              "originalUrl": "https://pbs.twimg.com/media/C22X9ODXgAABGKS.jpg",
              "squarePreviewUrl": "https://pbs.twimg.com/media/C22X9ODXgAABGKS.jpg",
              "tags": Array [],
            },
          ],
          "location": Object {
            "lat": 42.359927299999995,
            "lng": -71.0576853,
          },
          "requestedAtString": "March 7, 2017, 12:59 PM",
          "service": Object {
            "name": "Cosmic Intervention",
          },
          "status": "closed",
          "statusNotes": "Found Thanos. Smashed him into the floor with all of us standing around.",
          "updatedAtString": "April 8, 2017, 12:59 PM",
        }
      }
      store={
        AppStore {
          "_liveAgentButtonId": "",
          "accessibility": Accessibility {
            "el": null,
            "messageListenerDisposer": null,
          },
          "addressSearch": AddressSearch {
            "searchPopupWidth": 0,
          },
          "allServices": AllServices {
            "groups": Array [
              Group {
                "description": "Pet Ownership, Problems with Animals, Rodents & Pests",
                "id": "Animals",
                "name": "Animals",
              },
              Group {
                "description": "NEEDS DESCRIPTION",
                "id": "Fees-Pricing",
                "name": "Fees and Pricing",
              },
              Group {
                "description": "Crime, Fire Prevention & Control, Healthy Living & Inspections",
                "id": "Health-Safety",
                "name": "Health, Safety, and Social Services",
              },
              Group {
                "description": "Park Maintenance, Permits, Tree Planting, Donations",
                "id": "Parks-Public",
                "name": "Parks and Public Space",
              },
              Group {
                "description": "Home Ownership, Renting, Inspections, Code Enforcement, Permitting, Complaints",
                "id": "Property-Construction",
                "name": "Property, Housing and Construction",
              },
              Group {
                "description": "NEEDS DESCRIPTION",
                "id": "Schools",
                "name": "Schools",
              },
              Group {
                "description": "Plowing, snow removal",
                "id": "Snow",
                "name": "Snow or Weather Related Concerns",
              },
              Group {
                "description": "Signs & Signals, Parking, Driving, Public Transportation, Taxi Issues, Road Maintenance",
                "id": "Streets-Sidewalks",
                "name": "Streets, Sidewalks, Traffic, and Vehicles",
              },
              Group {
                "description": "Trash, Recycling, Graffiti, Dumping, Street Cleaning, Neighborhood Cleanup",
                "id": "Trash-Graffiti",
                "name": "Trash, Recycling, and Graffiti",
              },
              Group {
                "description": "NEEDS DESCRIPTION",
                "id": "Utilities",
                "name": "Utilities",
              },
              Group {
                "description": "General service requests that don’t fit in another category",
                "id": "Other",
                "name": "Other Services",
              },
            ],
            "otherGroup": Group {
              "description": "These services have not yet been categorized",
              "id": "",
              "name": "Uncategorized Services",
            },
          },
          "apiKeys": Object {},
          "browserLocation": BrowserLocation {
            "watchId": null,
          },
          "languages": Array [],
          "requestSearch": RequestSearch {
            "boundsSet": false,
            "boundsSetDisposer": null,
            "debouncedSearch": [Function],
            "mapMoved": false,
            "searchDisposer": null,
          },
          "ui": Ui {
            "handleResize": [Function],
            "handleScroll": [Function],
          },
        }
      }
      submitted={true}
    />
  </FormDialog>
</div>
`;

exports[`rendering contact 1`] = `
<div
  className={
    Object {
      "data-css-1prrpm3": "",
    }
  }
>
  <div
    className="p-a300 p-a800--xl br br-t400 br--y css-1aypqv0"
    style={
      Object {
        "maxWidth": 800,
      }
    }
  >
    <form
      onSubmit={[Function]}
    >
      <div>
        <div
          className=" sh sh--y "
        >
          <h1
            className="sh-title"
          >
            Cosmic Incursion
          </h1>
        </div>
        <p
          className="m-v300 t--info"
        >
          Why are we asking for your contact information? We’ll use it to email you about the status of your report. We may also have to follow up with you.
        </p>
        <p
          className="m-v300 t--subinfo"
        >
          <strong>
            We won’t make your information public.
          </strong>
           
        </p>
        <div
          className="txt"
        >
          <label
            className="txt-l"
            htmlFor="ContactPane-firstName"
          >
            First Name 
            <span
              className="t--req"
            >
              Required
            </span>
          </label>
          <input
            aria-required={true}
            className="txt-f"
            id="ContactPane-firstName"
            name="firstName"
            onChange={[Function]}
            placeholder="First Name"
            style={
              Object {
                "width": "100%",
              }
            }
            type="text"
            value=""
          />
        </div>
        <div
          className="txt"
        >
          <label
            className="txt-l"
            htmlFor="ContactPane-lastName"
          >
            Last Name 
            <span
              className="t--req"
            >
              Required
            </span>
          </label>
          <input
            aria-required={true}
            className="txt-f"
            id="ContactPane-lastName"
            name="lastName"
            onChange={[Function]}
            placeholder="Last Name"
            style={
              Object {
                "width": "100%",
              }
            }
            type="text"
            value=""
          />
        </div>
        <div
          className="txt"
        >
          <label
            className="txt-l"
            htmlFor="ContactPane-email"
          >
            Email 
            <span
              className="t--req"
            >
              Required
            </span>
          </label>
          <input
            aria-required={true}
            className="txt-f"
            id="ContactPane-email"
            name="email"
            onChange={[Function]}
            placeholder="Email"
            style={
              Object {
                "width": "100%",
              }
            }
            type="email"
            value=""
          />
        </div>
        <div
          className="txt"
        >
          <label
            className="txt-l"
            htmlFor="ContactPane-phone"
          >
            Phone
          </label>
          <input
            className="txt-f"
            id="ContactPane-phone"
            name="phone"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onPaste={[Function]}
            placeholder="Phone"
            style={
              Object {
                "width": "100%",
              }
            }
            type="tel"
            value=""
          />
        </div>
        <div
          className="m-v500 g"
        >
          <div
            className="cb"
          >
            <input
              checked={false}
              className="cb-f"
              id="ContactPane-remember"
              name="remember"
              onChange={[Function]}
              type="checkbox"
              value="true"
            />
            <label
              className="cb-l"
              htmlFor="ContactPane-remember"
            >
              Remember contact info on this computer
            </label>
          </div>
        </div>
      </div>
      <div
        className="g m-v500 css-1k8tu3j"
      >
        <div
          className="g--8 t--info m-v200 css-hn0yet"
        >
          <span>
            Please fill out 
            <span
              className="t--req"
            >
              required
            </span>
             fields to continue
          </span>
        </div>
        <button
          className="btn g--4"
          disabled={true}
          type="submit"
        >
          Submit Request
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`rendering location 1`] = `
<div
  className={
    Object {
      "data-css-k71z4": "",
    }
  }
>
  <div
    className=" br br-t400 br--y css-1aypqv0"
    style={Object {}}
  >
    <div
      className={
        Object {
          "data-css-xmixgj": "",
        }
      }
    >
      <div
        className="css-m89tpz"
      >
        <div
          className="p-a300"
        >
          <form
            action="#"
            className="sf sf--md sf--y"
            onSubmit={[Function]}
          >
            <div
              className="sf-i"
            >
              <input
                aria-label="Address search field"
                className="sf-i-f"
                onChange={[Function]}
                placeholder="Search address…"
                type="text"
                value=""
              />
              <button
                className="sf-i-b"
                disabled={true}
                type="submit"
              >
                Search
              </button>
            </div>
          </form>
        </div>
        <div
          className={
            Object {
              "data-css-129qykh": "",
            }
          }
        />
        <div
          className="g p-a300 css-1k8tu3j"
        >
          <div
            className="g--7 t--subinfo m-v200"
          >
            <a
              href="javascript:void(0)"
              onClick={[Function]}
            >
              Continue without location
            </a>
          </div>
          <button
            className="btn g--5"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Next
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`rendering questions 1`] = `
<div
  className={
    Object {
      "data-css-1prrpm3": "",
    }
  }
>
  <div
    className="p-a300 p-a800--xl br br-t400 br--y css-1aypqv0"
    style={Object {}}
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className=" sh sh--y "
      >
        <h1
          className="sh-title"
        >
          Cosmic Incursion
        </h1>
      </div>
      <div
        className="t--intro m-v300"
      >
        Bad things getting in from other universes
      </div>
      <div
        className="g g--top"
      >
        <div
          className="g--7 m-v500"
        >
          <div
            className="txt"
          >
            <label
              className="txt-l"
              htmlFor="QuestionsPane-description"
              style={
                Object {
                  "marginTop": 0,
                }
              }
            >
              Description
            </label>
            <textarea
              className="txt-f"
              id="QuestionsPane-description"
              name="description"
              onChange={[Function]}
              placeholder="Please describe your request"
              rows="5"
              value=""
            />
          </div>
          <div>
            <div
              className="txt"
            >
              <label
                className="txt-l"
                htmlFor="ST-CMTS"
              >
                Please provide any other relevant information:
                 
              </label>
              <textarea
                aria-required={false}
                className="txt-f"
                id="ST-CMTS"
                name="ST-CMTS"
                onChange={[Function]}
                rows="5"
                value=""
              />
            </div>
          </div>
          <hr
            aria-hidden={true}
            className="hr hr--dash m-v500"
          />
          <div>
            <div
              className="txt"
            >
              <label
                className="txt-l"
                htmlFor="INFO-CSIRMV1"
              >
                **All cosmic incursion cases should be followed up with a phone call to Alpha Flight.**
                 
              </label>
              <input
                aria-required={false}
                className="txt-f "
                id="INFO-CSIRMV1"
                name="INFO-CSIRMV1"
                onChange={[Function]}
                type="text"
                value=""
              />
            </div>
          </div>
          <hr
            aria-hidden={true}
            className="hr hr--dash m-v500"
          />
          <div>
            <div
              aria-labelledby="SR-CSIRMV1-label"
              role="radiogroup"
            >
              <div
                className="m-v300"
              >
                <label
                  className="txt-l"
                  id="SR-CSIRMV1-label"
                >
                  How many dimensions have been breached?
                   
                  <span
                    className="t--req"
                  >
                    Required
                  </span>
                </label>
              </div>
              <div
                className="g"
              >
                <div
                  className="g--12"
                >
                  <div
                    className="ra"
                  >
                    <input
                      checked={false}
                      className="ra-f"
                      id="SR-CSIRMV1-One"
                      name="SR-CSIRMV1"
                      onChange={[Function]}
                      type="radio"
                      value="One"
                    />
                    <label
                      className="ra-l"
                      htmlFor="SR-CSIRMV1-One"
                    >
                      One
                    </label>
                  </div>
                  <div
                    className="ra"
                  >
                    <input
                      checked={false}
                      className="ra-f"
                      id="SR-CSIRMV1-Two"
                      name="SR-CSIRMV1"
                      onChange={[Function]}
                      type="radio"
                      value="Two"
                    />
                    <label
                      className="ra-l"
                      htmlFor="SR-CSIRMV1-Two"
                    >
                      Two
                    </label>
                  </div>
                  <div
                    className="ra"
                  >
                    <input
                      checked={false}
                      className="ra-f"
                      id="SR-CSIRMV1-Three"
                      name="SR-CSIRMV1"
                      onChange={[Function]}
                      type="radio"
                      value="Three"
                    />
                    <label
                      className="ra-l"
                      htmlFor="SR-CSIRMV1-Three"
                    >
                      Three
                    </label>
                  </div>
                  <div
                    className="ra"
                  >
                    <input
                      checked={false}
                      className="ra-f"
                      id="SR-CSIRMV1-More than Three"
                      name="SR-CSIRMV1"
                      onChange={[Function]}
                      type="radio"
                      value="More than Three"
                    />
                    <label
                      className="ra-l"
                      htmlFor="SR-CSIRMV1-More than Three"
                    >
                      More than Three
                    </label>
                  </div>
                  <span />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="g--5 m-v500"
        >
          <div
            className="br br-a200"
          >
            <div
              aria-disabled={false}
              className="css-19ub58"
              onClick={[Function]}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDragStart={[Function]}
              onDrop={[Function]}
              style={Object {}}
            >
              <img
                alt=""
                aria-hidden={true}
                height="324"
                src="/assets/img/image-upload.png"
                style={
                  Object {
                    "display": "block",
                    "height": "auto",
                    "width": "100%",
                  }
                }
                width="479"
              />
              <input
                accept="image/*"
                autoComplete="off"
                disabled={false}
                multiple={false}
                onChange={[Function]}
                style={
                  Object {
                    "display": "none",
                  }
                }
                type="file"
              />
            </div>
            <div
              className="br br-t200"
              style={
                Object {
                  "position": "relative",
                }
              }
            >
              <button
                className="btn btn--w btn--b"
                onClick={[Function]}
                style={
                  Object {
                    "visibility": "visible",
                    "width": "100%",
                  }
                }
                type="button"
              >
                Upload Photo
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        className="g m-v500 css-1k8tu3j"
      >
        <div
          className="g--9 t--info m-v200 css-hn0yet"
        >
          <span>
            Please fill out 
            <span
              className="t--req"
            >
              required
            </span>
             fields to continue
          </span>
        </div>
        <button
          className="btn g--3"
          disabled={true}
          type="submit"
        >
          Next
        </button>
      </div>
    </form>
  </div>
</div>
`;
